README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ShrinkCovMat: Shrinkage Covariance Matrix Estimators

[![Github
version](https://img.shields.io/badge/GitHub%20-1.4.2-green.svg)](%22commits/master%22)
[![Travis-CI Build
Status](https://travis-ci.org/AnestisTouloumis/ShrinkCovMat.svg?branch=master)](https://travis-ci.org/AnestisTouloumis/ShrinkCovMat)
[![Project Status: Active The project has reached a stable, usable state
and is being actively
developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![Codecov test
coverage](https://codecov.io/gh/AnestisTouloumis/ShrinkCovMat/branch/master/graph/badge.svg)](https://codecov.io/gh/AnestisTouloumis/ShrinkCovMat?branch=master)

[![CRAN
Version](https://www.r-pkg.org/badges/version/ShrinkCovMat?color=blue)](https://CRAN.R-project.org/package=ShrinkCovMat)
[![CRAN
Downloads](https://cranlogs.r-pkg.org/badges/grand-total/ShrinkCovMat?color=blue)](https://cranlogs.r-pkg.org/badges/grand-total/ShrinkCovMat)
[![CRAN
Downloads](https://cranlogs.r-pkg.org/badges/ShrinkCovMat)](https://CRAN.R-project.org/package=ShrinkCovMat)

## Installation

You can install the release version of `ShrinkCovMat`:

``` r
install.packages("ShrinkCovMat")
```

The source code for the release version of `ShrinkCovMat` is available
on CRAN at:

  - <https://CRAN.R-project.org/package=ShrinkCovMat>

Or you can install the development version of `ShrinkCovMat`:

``` r
# install.packages('devtools')
devtools::install_github("AnestisTouloumis/ShrinkCovMat")
```

The source code for the development version of `ShrinkCovMat` is
available on github at:

  - <https://github.com/AnestisTouloumis/ShrinkCovMat>

To use `ShrinkCovMat`, you should first load the package as follows:

``` r
library(ShrinkCovMat)
```

## Usage

This package provides the nonparametric Stein-type shrinkage covariance
matrix estimators proposed by Touloumis (2015). These estimators are
suitable and statistically efficient regardless of the dimensionality.

Each of the implemented shrinkage covariance matrix estimators is a
convex linear combination of the sample covariance matrix and of a
target matrix. Three options are considered for the target matrix:

  - the identity matrix (`shrinkcovmat.identity`),
  - the scaled identity matrix (`shrinkcovmat.equal`),
  - the diagonal matrix with diagonal elements the corresponding sample
    variances (`shrinkcovmat.unequal`).

Estimation of the corresponding optimal shrinkage intensities is
discussed in Touloumis (2015).

The utility function `targetselection` is designed to ease the selection
of the target matrix.

## Example

Consider the colon cancer data example analyzed in Touloumis (2015). The
data consists of two tissue groups: the normal tissue group and the
tumor tissue group.

``` r
data(colon)
normal_group <- colon[, 1:40]
tumor_group <- colon[, 41:62]
```

To decide the target matrix for covariance matrix of the normal group,
inspect the output of the function `targetselection`:

``` r
targetselection(normal_group)
#> OPTIMAL SHRINKAGE INTENSITIES FOR THE TARGET MATRIX WITH 
#> Equal variances   : 0.1401 
#> Unit variances    : 0.1125 
#> Unequal variances : 0.14 
#> 
#> SAMPLE VARIANCES 
#> Range   : 0.4714 
#> Average : 0.0882
```

The estimated optimal shrinkage intensity for the scaled identity matrix
is slightly larger than the other two and the sample variances appear to
be of similar magnitude. Thus, the scaled identity matrix seems to be
the most appropriate target for the covariance estimated. The resulting
covariance matrix estimator is:

``` r
estimated_covariance_normal <- shrinkcovmat.equal(normal_group)
estimated_covariance_normal
#> SHRINKAGE ESTIMATION OF THE COVARIANCE MATRIX 
#> 
#> Estimated Optimal Shrinkage Intensity = 0.1401 
#> 
#> Estimated Covariance Matrix [1:5,1:5] =
#>        [,1]   [,2]   [,3]   [,4]   [,5]
#> [1,] 0.0396 0.0107 0.0101 0.0214 0.0175
#> [2,] 0.0107 0.0499 0.0368 0.0171 0.0040
#> [3,] 0.0101 0.0368 0.0499 0.0147 0.0045
#> [4,] 0.0214 0.0171 0.0147 0.0523 0.0091
#> [5,] 0.0175 0.0040 0.0045 0.0091 0.0483
#> 
#> Target Matrix [1:5,1:5] =
#>        [,1]   [,2]   [,3]   [,4]   [,5]
#> [1,] 0.0882 0.0000 0.0000 0.0000 0.0000
#> [2,] 0.0000 0.0882 0.0000 0.0000 0.0000
#> [3,] 0.0000 0.0000 0.0882 0.0000 0.0000
#> [4,] 0.0000 0.0000 0.0000 0.0882 0.0000
#> [5,] 0.0000 0.0000 0.0000 0.0000 0.0882
```

We follow a similar procedure for the tumor group:

``` r
targetselection(tumor_group)
#> OPTIMAL SHRINKAGE INTENSITIES FOR THE TARGET MATRIX WITH 
#> Equal variances   : 0.1956 
#> Unit variances    : 0.1705 
#> Unequal variances : 0.1955 
#> 
#> SAMPLE VARIANCES 
#> Range   : 0.4226 
#> Average : 0.0958
```

As before, we may conclude that the scaled identity matrix seems to be
the most suitable target matrix. The resulting covariance matrix
estimate for the tumor group is:

``` r
estimated_covariance_tumor <- shrinkcovmat.equal(tumor_group)
estimated_covariance_tumor
#> SHRINKAGE ESTIMATION OF THE COVARIANCE MATRIX 
#> 
#> Estimated Optimal Shrinkage Intensity = 0.1956 
#> 
#> Estimated Covariance Matrix [1:5,1:5] =
#>        [,1]   [,2]   [,3]   [,4]   [,5]
#> [1,] 0.0490 0.0179 0.0170 0.0195 0.0052
#> [2,] 0.0179 0.0450 0.0265 0.0092 0.0034
#> [3,] 0.0170 0.0265 0.0465 0.0084 0.0031
#> [4,] 0.0195 0.0092 0.0084 0.0498 0.0036
#> [5,] 0.0052 0.0034 0.0031 0.0036 0.0361
#> 
#> Target Matrix [1:5,1:5] =
#>        [,1]   [,2]   [,3]   [,4]   [,5]
#> [1,] 0.0958 0.0000 0.0000 0.0000 0.0000
#> [2,] 0.0000 0.0958 0.0000 0.0000 0.0000
#> [3,] 0.0000 0.0000 0.0958 0.0000 0.0000
#> [4,] 0.0000 0.0000 0.0000 0.0958 0.0000
#> [5,] 0.0000 0.0000 0.0000 0.0000 0.0958
```

## How to cite

``` 

To cite the R package 'ShrinkCovMat' in publications, please use:

  Touloumis, A. (2015) Nonparametric Stein-type Shrinkage
  Covariance Matrix Estimators in High-Dimensional Settings,
  Computational Statistics & Data Analysis 83, 251-261.

A BibTeX entry for LaTeX users is

  @Article{,
    title = {Nonparametric Stein-type Shrinkage Covariance Matrix Estimators in High-Dimensional Settings},
    author = {{Anestis Touloumis}},
    year = {2015},
    journal = {Computational Statistics & Data Analysis},
    volume = {83},
    pages = {251--261},
  }
```

# References

<div id="refs" class="references">

<div id="ref-Touloumis2015">

Touloumis, A. (2015) Nonparametric Stein-type Shrinkage Covariance
Matrix Estimators in High-Dimensional Settings. *Computational
Statistics & Data Analysis*, **83**, 251–261.

</div>

</div>


inst/NEWS
Version 1.4.1 [2019-10-31]
 * Fixed ORCID in DESCRIPTION.

Version 1.4.0 [2019-07-30]
 * Complied with CRAN Solaris requirements.
 * Re-submitted to CRAN.
 
Version 1.3.0 [2019-07-25]
 * Released on CRAN.

Version 1.2.15 [2019-07-24]
 * Cosmetic changes in the R files.
 
Version 1.2.14 [2019-07-23]
 * Cosmetic changes in unit test files.

Version 1.2.13 [2019-07-22]
 * Added a unit test for target selection.
 * Created an S3 class for target selection.
 
Version 1.2.12 [2019-07-14]
 * Improved readability of R code.
 
Version 1.2.11 [2019-07-13]
 * Improved readability of R code.
 
Version 1.2.10 [2019-07-10]
 * Added more unit-tests.
 
Version 1.2.9 [2019-07-09]
 * Cleared lintr issues.
 * Cosmetic changes in the examples and README.Rmd.

Version 1.2.8 [2019-07-02]
 * Removed lintr dependency.
 * Updated .travis.yml.
 
Version 1.2.7 [2019-07-02]
 * Added lint unit-testing.
 * Moving to snake_case_style.
 * Added lintr dependency.

Version 1.2.6 [2019-06-11]
 * Minor changes in the code.
 
Version 1.2.5 [2019-06-06]
 * Removed code dependencies with sample covariance matrix.
 
Version 1.2.4 [2019-06-04]
 * Added Rcpp functionality.

Version 1.2.3 [2019-05-30]
 * Added code coverage.
 
Version 1.2.2 [2019-05-29]
 * Fixed a bug thanks to Øyvind Bleka. 
 * Introduced unit testing.

Version 1.2.1 [2017-08-21]
 * Updated documentation using roxygen2.

Version 1.2.0 [2017-07-11]
 * Finalized README.Rmd.
 * Released on CRAN.
 
Version 1.1.9 [2017-07-10]
 * Updated README.Rmd.

Version 1.1.8 [2017-07-08]
 * Updated documentation.

Version 1.1.7 [2017-07-07]
 * Modified README.Rmd.
 
Version 1.1.6 [2017-06-21]
 * Modified .gitignore and .Rbuildignore.

Version 1.1.5 [2017-06-20]
 * Updated README.Rmd.
 * Updated Rcode - reference style.
 
Version 1.1.4 [2017-06-15]
 * Added README.Rmd.
 * Updated documentation - reference style.

Version 1.1.3 [2017-06-12]
 * Updated DESCRIPTION- links to the GitHub site.

Version 1.1.2 [2016-05-22]
 * Fixed a bug in the function shrinkcovmat.unequal 
   (reported by Prof. Thomas A. Severini).
 * Updated documentation.
 * Updated R files.
 * Updated CITATION.

Version 1.1.1 [2015-09-08]
 * Updated email address of the maintainer.

Version 1.1.0 [2015-01-08]
 * Updated CITATION.

Version 1.0.3 [2014-11-02]
 * Updated CITATION.
 * Updated documention.

Version 1.0.2 [2014-10-20]
 * Added function targetselection. 
 * Updated CITATION.
 * Updated documention.

Version 1.0.1 [2014-08-12]
 * Updated documentation. 
 * Created CITATION. 

Version 1.0.0 [2014-04-17]
 * The first version of the R package ShrinkCovMat is released on CRAN.
 

R/targetselection.R
#' Target Matrix Selection
#'
#' Implements the rule of thumb proposed by Touloumis (2015) for target matrix
#' selection. If the estimated optimal shrinkage intensities of the three
#' target matrices are of similar magnitude, then the average and the range of
#' the sample variances should be inspected in order to adopt the most
#' plausible target matrix.
#'
#' The rows of the data matrix \code{data} correspond to variables and the
#' columns to subjects.
#'
#' @param data a numeric matrix containing the data.
#' @param centered a logical indicating if the mean vector is the zero vector.
#' @return Prints the estimated optimal shrinkage intensities, the range and
#' average of the sample variances and returns an object of the class
#' 'targetsel' that has components: \item{optimal_sphericity}{The estimated
#' optimal intensity for a target matrix with equal variances.}
#' \item{optimal_identity}{The estimated optimal shrinkage
#' intensity for the identity target matrix.} \item{optimal_diagonal}{The
#' estimated optimal intensity for a target matrix with unequal variances.}
#' \item{range}{The range of the sample variances.} \item{average}{The average
#' of the sample variances.}
#' @author Anestis Touloumis
#' @references Touloumis, A. (2015) Nonparametric Stein-type Shrinkage
#' Covariance Matrix Estimators in High-Dimensional Settings.
#' \emph{Computational Statistics & Data Analysis} \bold{83}, 251--261.
#' @examples
#' data(colon)
#' normal_group <- colon[, 1:40]
#' targetselection(normal_group)
#' ## Similar intensities, the range of the sample variances is small and the
#' ## average is not close to one. The scaled identity matrix seems to be the
#' ## most suitable target matrix for the normal group.
#'
#' tumor_group <- colon[, 41:62]
#' targetselection(tumor_group)
#' ## Similar intensities, the range of the sample variances is small and the
#' ## average is not close to one. The scaled identity matrix seems to be the
#' ## most suitable target matrix for the colon group.
#' @export
targetselection <- function(data, centered = FALSE) {
  if (!is.matrix(data)) data <- as.matrix(data)
  p <- nrow(data)
  n <- ncol(data)
  if (!centered) {
    if (n < 4) stop("The number of columns should be greater than 3")
    sample_variances <- apply(data, 1, var)
    trace_statistics <- trace_stats_uncentered(data) # nolintr
    trace_sigma_hat <- trace_statistics[1]
    trace_sigma_squared_hat <- trace_statistics[2]
    lambda_hat_sphericity <- (trace_sigma_hat^2 + trace_sigma_squared_hat) /
      (n * trace_sigma_squared_hat + (p - n + 1) / p * trace_sigma_hat^2)
    lambda_hat_sphericity <- min(lambda_hat_sphericity, 1)
    lambda_hat_identity <- (trace_sigma_hat^2 + trace_sigma_squared_hat) /
      (n * trace_sigma_squared_hat + trace_sigma_hat^2 -
        2 * trace_sigma_hat * (n - 1) + p * (n - 1))
    lambda_hat_identity <- max(0, min(lambda_hat_identity, 1))
    trace_diagonal_sigma_sq_hat <- trace_statistics[3]
    lambda_hat_diagonal <- (trace_sigma_hat^2 + trace_sigma_squared_hat -
      (2 - 2 / n) * trace_diagonal_sigma_sq_hat) /
      (n * trace_sigma_squared_hat + trace_sigma_hat^2 -
        (n + 1 - 2 / n) * trace_diagonal_sigma_sq_hat)
    lambda_hat_diagonal <- max(0, min(lambda_hat_diagonal, 1))
  } else {
    if (n < 2) stop("The number of columns should be greater than 1")
    sample_variances <- apply(data, 1, function(x) mean(x^2))
    trace_statistics <- trace_stats_centered(data) # nolintr
    trace_sigma_hat <- trace_statistics[1]
    trace_sigma_squared_hat <- trace_statistics[2]
    trace_diagonal_sigma_sq_hat <- trace_statistics[3]
    lambda_hat_sphericity <- (trace_sigma_hat^2 + trace_sigma_squared_hat) /
      ((n + 1) * trace_sigma_squared_hat + (p - n) / p * trace_sigma_hat^2)
    lambda_hat_sphericity <- min(lambda_hat_sphericity, 1)
    lambda_hat_identity <- (trace_sigma_hat^2 + trace_sigma_squared_hat) /
      ((n + 1) * trace_sigma_squared_hat + trace_sigma_hat^2 -
        2 * trace_sigma_hat * n + p * n)
    lambda_hat_identity <- max(0, min(lambda_hat_identity, 1))
    lambda_hat_diagonal <- (trace_sigma_hat^2 + trace_sigma_squared_hat -
      (2 - 2 / (n + 1)) * trace_diagonal_sigma_sq_hat) /
      ((n + 1) * trace_sigma_squared_hat + trace_sigma_hat^2 -
        (n + 2 - 2 / (n + 1)) * trace_diagonal_sigma_sq_hat)
    lambda_hat_diagonal <- max(0, min(lambda_hat_diagonal, 1))
  }
  ans <- list(
    optimal_sphericity = lambda_hat_sphericity,
    optimal_identity = lambda_hat_identity,
    optimal_diagonal = lambda_hat_diagonal,
    range = abs(diff(range(sample_variances))),
    average = mean(sample_variances)
  )
  class(ans) <- "targetsel"
  ans
}
